### ISSUE TITLE
Idea: Separate Image runConfig from file system layers

### ISSUE THREAD

**Issue by jlhawn** [docker/docker#9037]
Currently, an Image ID maps directly to a file system layer. It also is directly associated with the container config which was used to commit that file system layer and the config to use when starting a new container using this image. This is not ideal for various reasons:

- binding an image's randomly generated ID to file system content rather than using content-addressable layers
- creating unnecessary file system layers for every step while building an image (typically, only ADD/COPY/RUN/ONBUILD Dockerfile commands will produce file system changes)
- Image history is constructed by following a chain of Parent IDs, requiring opening and reading a separate file for every layer.

Here is the current image format as stored by Docker:

```
type Image struct {
	ID              string
	Parent          string
	Comment         string
	Created         time.Time
	Container       string
	ContainerConfig runconfig.Config
	DockerVersion   string
	Author          string
	Config          *runconfig.Config
	Architecture    string
	OS              string
	Size            int64
}
```

While I do not have any formal proposal yet, I've discussed various ideas with project maintainers and other members of the #distribution team. So far, our best idea is to separate the current `image.Image` struct type into 2 structs:

```
type Image struct {
	name struct {
		repository string
		tag        string
	}

	Created        time.Time
	DockerVersion  string
	Maintainers    []string
	Architecture   string
	OS             string

	Config         *runconfig.Config

	FSLayers []string  // Tarsums of file system layers
	History  []string
}

type FileSystemSnapshot struct {
	ID              string  // Randomly Generated IDs which are specific
	Parent          string  // to the storage driver on the local host

	Checksum  	   string
	ContentAddress string

	Created         time.Time
	ContainerConfig runconfig.Config
	DockerVersion   string

	VirtualSize     int64
	LayerDiffSize   int64
}
```

The point of this issue is not to propose a solution, but to discuss issues with the current image format and discuss ideas for an alternative, next generation format to be used by the docker engine.

**Comment by LK4D4** [comment]
@tonistiigi did something changed for this issue?

**Comment by tonistiigi** [comment]
@LK4D4 Yes all the problem points in the original issue were addressed in content addressability in v1.10
